source/common/common/inline_map_registry.h

```diff
@@ -0,0 +1,389 @@
+#pragma once
+
+#include <cstddef>
+#include <cstdint>
+
+#include "source/common/common/assert.h"
+#include "source/common/common/macros.h"
+#include "source/common/common/utility.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/node_hash_set.h"
+
+namespace Envoy {
+
+/**
+ * This class could be used as an alternative of normal hash map to store the key/value pairs.
+ * But by this class, you can register some frequently used keys as inline keys and get the
+ * handle for the key. Then you can use the handle to access the key/value pair in the map without
+ * even one time hash searching.
+ *
+ * This is useful when some keys are frequently used and the keys are known at compile time or
+ * bootstrap time. You can get superior performance by using the inline handle. For example, the
+ * filter state always uses the filter name as the key and the filter name is known at compile time.
+ * By using the inline handle, the filter state could get the key/value pair without any hash.
+ *
+ * This class also provides the interface to access the key/value pair by the normal key and the
+ * interface has similar searching performance as the normal hash map. But the insertion and removal
+ * by the normal key is slower than the normal hash map.
+ *
+ *
+ * These is usage example:
+ *
+ *   // Define registry to store the inline keys. The registry should have lifetime that no shorter
+ *   // than all inline maps created by the registry.
+ *   InlineMapRegistry registry;
+ *
+ *   // Register the inline key. We should never do this after bootstrapping. Typically, we can
```

Comment:
Http::HeaderMap uses CONSTRUCT_ON_FIRST_USE to hold structures with predefined keys, I think. Not a global variable per key.

Reply:
I think only all well known keys (like method, authority, content-length) are registered when the registry is finalized and all related inline keys are stored in a struct object.
But other custom inline keys are global variables per key?

Reply:
I hope not; they need to do this under a lazy-init macro otherwise there's a fiasco.

Reply:
As a library, the `InlineMapRegistry` and `InlineMap` will not force users to use this way.

But this way is widely used in the envoy. If you search `Http::RegisterCustomInlineHeader<Http::CustomInlineHeaderRegistry::Type::RequestHeaders>`, you will find lots related code. orz

As a library, we should ensure even users do that, everything will work as expected.